interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,956 @@
+// permission for the interface associated with the interface DB.
+type interfaceDB struct {
+	// interfaceDB contains a map from permission to permissionDB for a particular interface
+	PerPermission map[string]*permissionDB
+}
+
+// snapDB stores a map from interface name to the DB of rules pertaining to that
+// interface for the snap associated with the snap DB.
+type snapDB struct {
+	// snapDB contains a map from interface to interfaceDB for a particular snap
+	PerInterface map[string]*interfaceDB
+}
+
+// userDB stores a map from snap name to the DB of rules pertaining to that
+// snap for the user associated with the user DB.
+type userDB struct {
+	// userDB contains a map from snap to snapDB for a particular user
+	PerSnap map[string]*snapDB
+}
+
+// RuleDB stores a mapping from rule ID to rule, and a tree of rule IDs
+// searchable by user, snap, interface, permission, and pattern variant.
+type RuleDB struct {
+	mutex     sync.RWMutex
+	maxIDMmap maxidmmap.MaxIDMmap
+	ids       map[prompting.IDType]int
+	rules     []*Rule
+	perUser   map[uint32]*userDB
+	// notifyRule is a closure which will be called to record a notice when a
+	// rule is added, patched, or removed.
+	notifyRule func(userID uint32, ruleID prompting.IDType, data map[string]string) error
+}
+
+// New creates a new rule database, loads existing rules from the database file,
+// and returns the populated database.
+//
+// The given notifyRule closure may be called before `New()` returns, if a
+// previously-saved rule has expired or if there are conflicts between rules.
+//
+// The given notifyRule closure will be called when a rule is added, modified,
+// expired, or removed. In order to guarantee the order of notices, notifyRule
+// is called with the prompt DB lock held, so it should not block for a
+// substantial amount of time (such as to lock and modify snapd state).
+func New(notifyRule func(userID uint32, ruleID prompting.IDType, data map[string]string) error) (*RuleDB, error) {
+	maxIDFilepath := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "request-rule-max-id")
+	maxIDMmap, err := maxidmmap.OpenMaxIDMmap(maxIDFilepath)
+	if err != nil {
+		return nil, err
+	}
+	rdb := &RuleDB{
+		maxIDMmap:  maxIDMmap,
+		ids:        make(map[prompting.IDType]int),
+		rules:      make([]*Rule, 0),
+		perUser:    make(map[uint32]*userDB),
+		notifyRule: notifyRule,
+	}
+	if err = rdb.load(); err != nil {
+		logger.Noticef("%v; using new empty rule database", err)
+	}
+	return rdb, nil
+}
+
+// load reads the stored rules from the database file and populates the
+// receiving rule database.
+//
+// Returns an error if an existing prompt DB cannot be loaded. An expired rule
+// or rule conflict does not cause an error. If saving the DB is required but
+// the save fails, returns an error.
+//
+// If an error is returned, the rule database is not modified.
+func (rdb *RuleDB) load() error {
+	target := rdb.dbpath()
+	f, err := os.Open(target)
+	if err != nil {
```

Comment:
prehaps we could simply return nil on `errors.Is(err, fs.ErrNotExist)` so that it isn't noisy?